nova/scheduler/weights/vm_hv_ratio.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2021 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+VmHvRatioWeigher. Weigh hosts by how well they fit the VM size.
+"""
+from oslo_log import log as logging
+
+import nova.conf
+from nova.scheduler.mixins import HypervisorSizeMixin
+from nova.scheduler import weights
+
+LOG = logging.getLogger(__name__)
+
+CONF = nova.conf.CONF
+
+
+class VmHvRatioWeigher(weights.BaseHostWeigher, HypervisorSizeMixin):
+    minval = 0
+
```

Comment:
I think, it would also make sense to set the `maxval = 1`, so the results are more deterministic, and depend not on the subset of hosts chosen from placement.

Reply:
Then we need to increase the weight-multiplier a lot. Because we will get values like '0.0104' to '0.0026' (8 GiB on 768 Gib vs. 3 TiB), which will not have any effect on the total weight unless we have a high multiplier. If we normalize without maxval though, we would get 1 and 0.25.

Reply:
Yes, I was kind of working with the assumption of using a large multiplier.
More or less, I was expecting that the normal weights will be normalised to `1`, but the `VmHvRatioWeigher` should give us results clearly larger than the sum of the other possible weights.
And since there is 512MiB flavour, the factor should then probably be around 10.000.000

But yes, I think you are right, it doesn't look good that we need to crank up the factor just because we have small flavours. 
So we might want to have some kind of normalisation across flavours, but (at least me) not across the dynamically chosen set of hosts.
So, that would simply mean following weight: `constant / hypervisor_ram_mb`, and I am wondering, why not?
We can only chose the host, and the flavour is fixed in the scheduling decision anyway.

Another problem (for a later time), we probably do not want to differentiate between a 768GiB hypervisor and a hypothetical 1TiB one.
So, maybe we could also go with `constant-hypervisor_ram_mb.bit_length()` or (`constant-)(hypervisor_ram_mb - tolerance).bit_length()`

With `constant=24` (For 16TiB hypervisor), and `tolerance=1`, I get the following weights for the hypervisors
| Up to TiB RAM/Hypervisor | Weight |
| --- | --- |
|  16 | 0  |
|  8 |  1 |
|  4 |  2  |
|  2  |  3 |
|  1  |  4 |

Do we need to differentiate between the hypervisors in those categories?
Or we might do away with all those formulas and put in there a table of weights directly.

Reply:
Okay, I understand.

Reply:
So ... should we remove the `maxval` then or do you think the current implementation should stay?

Reply:
I think, the current implementation should stay.